File: model/src/main/java/model/AdminData.java
package model;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.time.YearMonth;
import java.util.HashMap;
import java.util.Map;

/**
 * Represents a definition of admin record for a reference period i.e. a year and month
 * Each unit is compose of an id and a map of key value pairs reprsenting the variables
 *
 */
public class AdminData {

    public static final String REFERENCE_PERIOD_FORMAT = "yyyyMM";

    @JsonProperty("period")
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = REFERENCE_PERIOD_FORMAT)
    private YearMonth referencePeriod;

    @JsonProperty("id")
    private String id;

    @JsonProperty("vars")
    private Map<String, String> variables;

    public AdminData(YearMonth referencePeriod, String id) {
        this.referencePeriod = referencePeriod;
        this.id = id;
        this.variables = new HashMap<>();
    }

    public YearMonth getReferencePeriod() {
        return referencePeriod;
    }

    public String getId() {
        return id;
    }

    public Map<String, String> getVariables() {
        return variables;
    }

    public void putVariable(String variable, String value) {
        this.variables.put(variable, value);
    }

}

File: model/src/main/scala/model/AdminData.scala
package model

import com.github.nscala_time.time.Imports.YearMonth
import play.api.libs.json.{ JsValue, Json, Writes }

/**
 * AdminData
 * ----------------
 * Author: haqa
 * Date: 23 October 2017 - 20:41
 * Copyright (c) 2017  Office for National Statistics
 */

case class AdminData(
    referencePeriod: YearMonth,
    id: String,
    //    `type`: AdminDataType,
    variables: Map[String, String] = Map()
) {
  def putVariable(variable: String, value: String): AdminData = this.copy(variables = this.variables ++
    Map(variable -> value))
}

object AdminData {

  final val REFERENCE_PERIOD_FORMAT = "yyyyMM"

  implicit val writer: Writes[AdminData] = new Writes[AdminData] {
    def writes(a: AdminData): JsValue = {
      Json.obj(
        "period" -> a.referencePeriod.toString(REFERENCE_PERIOD_FORMAT),
        "id" -> a.id,
        //        "type" -> a.`type`,
        "vars" -> a.variables
      )
    }
  }

}
